pkg/metrics/metrics.go

```diff
@@ -126,27 +192,76 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
 		opMgr.cache.Delete(op)
 		return
 	}
-	strStatus := opStatusUnknown
+	strStatus := string(SnapshotStatusTypeUnknown)
 	if status != nil {
 		strStatus = status.String()
 	}
-	duration := time.Since(ts).Seconds()
-	opMgr.opLatencyMetrics.WithLabelValues(op.Driver, op.Name, strStatus).Observe(duration)
+
+	// record invidual reconciliation time
+	reconciliationDuration := time.Since(ts).Seconds()
+	opMgr.opReconciliationLatencyMetrics.WithLabelValues(op.Driver, op.Name, op.SnapshotType, strStatus).Observe(reconciliationDuration)
+
+	// record total operation durations if operation is successful
+	if strStatus == string(SnapshotStatusTypeSuccess) {
+		switch op.Name {
+		case CreateSnapshotOperationName:
+			// time since user first created volumesnapshot until a volumesnapshot.Status.CreationTime is aded
+			operationDuration := time.Since(snapshot.CreationTimestamp.Time).Seconds()
+			opMgr.opRequestLatencyMetrics.WithLabelValues(op.Driver, op.Name, op.SnapshotType, strStatus).Observe(operationDuration)
+
+		case CreateSnapshotAndReadyOperationName:
+			// time since user first created volumesnapshot until ReadyToUse is true
+			operationDuration := time.Since(snapshot.CreationTimestamp.Time).Seconds()
+			opMgr.opRequestLatencyMetrics.WithLabelValues(op.Driver, op.Name, op.SnapshotType, strStatus).Observe(operationDuration)
+
+		case DeleteSnapshotOperationName:
+			// check if this is a cancellation. It is a cancellation if the snapshot deletion finished without a status set
+			// or there is a status set with ReadyToUse set to false.
+			if snapshot.Status == nil || (snapshot.Status.ReadyToUse != nil && !(*snapshot.Status.ReadyToUse)) {
+				// get duration from creation time to delete/cancellation
+				cancelOperationDuration := time.Since(snapshot.CreationTimestamp.Time).Seconds()
+
+				// emit this metric as a custom cancellation
+				opMgr.opRequestLatencyMetrics.WithLabelValues(
+					op.Driver,
+					CreateSnapshotAndReadyOperationName,
```

Comment:
Is there a cancel for CreateSnapshot?

Reply:
Added this

Reply:
not seeing it?

Reply:
Ah sorry, thought I did. Added now above